aten/src/ATen/FunctionalTensorWrapper.cpp

```diff
@@ -322,6 +322,57 @@ void sync(const c10::List<c10::optional<Tensor>> t_list) {
   }
 }
 
+bool isFunctionalTensor(const at::Tensor& tensor) {
+  return tensor.unsafeGetTensorImpl()->key_set().has(c10::DispatchKey::Functionalize);
+}
+
+TORCH_API bool isFunctionalTensor(const c10::optional<Tensor>& t) {
+  if (t.has_value()) {
+    return isFunctionalTensor(*t);
+  } else {
+    return false;
+  }
+}
+
+TORCH_API bool isFunctionalTensor(const c10::List<Tensor>& t_list) {
```

Comment:
The 3 functions on list are exactly the same. Light templating and auto should allow you to merge them no?

Reply:
I made a modest attempt at this, but I think it's a little annoying to do - I can't use a generic template for it, since otherwise stuff like `c10::impl::ListElementReference<c10::optional<at::Tensor>>` would incorrectly try to match against it (before, we would implicitly convert to `c10::optional<Tensor>` and call that version).

I think I could fix that using `std::enable_if` to specialize the template for list-like types, but I kind of felt like that's more confusing to look at than just stamping out 3 copies of the 5 line function.